### DISCUSSION TITLE
Use pkg with MacBook Pro M3

### DISCUSSION THREAD

**Discussion by Drilmo** [yao-pkg/pkg#58] [Q&A] [answered]
Hello, I have several questions about pkg on MacBook Pro M3:

Is it possible to use it with pkg? 
And if so, I'd like to know if it's possible to create .exe files?

Thanks in advance for your feedback

**Comment by Drilmo** [comment]
I'd like to point out that I've tried to migrate a Windows project and I get the following error:
`Error! Not able to build for 'win' here, only for 'macOS'`

The windows project is a CLI application for some Windows colleagues, this application was packaged by pkg. Except that for other needs I've switched to Mac and I don't see how I can simply provide them without going through pkg. But is it feasible?

**Reply by robertsLando** [reply]
It depends if your module is using native node modules or not

**Reply by Drilmo** [reply]
Thanks for your answer, 
what do you mean by "native node modules" ? What would allow me to answer you?

Here is my package.json :

```json
{
  "name": "xxxxx",
  "version": "1.9.0",
  "description": "",
  "main": "dist/index.js",
  "bin": {
    "xxxx": "dist/index.js"
  },
  "scripts": {
    "start": "pnpm build && node dist/",
    "prewatch": "pnpm prebuild ",
    "preinstall": "npx only-allow pnpm",
    "watch": "pnpm tsc-watch --onSuccess \"node dist/\" ",
    "prebuild": "node -p \"'export const VERSION = ' + JSON.stringify(require('./package.json').version) + ';'\" > src/version.ts",
    "build": "pnpm rimraf dist/ && pnpm tsc",
    "test": "jest --config=./src/jest.config.ts",
    "createExe": "pnpm build && pnpm rimraf executable/v%npm_package_version% && pkg . ",
    "packit" : "pkg . --no-bytecode --public --public-package '*'",
    "newVersion": "pnpm createExe && pnpm auto-changelog && git add . && git commit -m \"chore: :bookmark: nouvelle version v%npm_package_version%\" && git tag v%npm_package_version% && git push && git push --tags",
    "installGlobalWithPnpm": "pnpm build && pnpm link --global",
    "uninstallGlobalWithPnpm": "pnpm uninstall --global %npm_package_name%",
    "auto-changelog": "pnpm rimraf CHANGELOG.md && auto-changelog -p --hide-credit "
  },
  "auto-changelog": {
    "commitUrl": "https://xxxxx/commits/{id}",
    "template": "changelog-template.hbs",
    "commitLimit": false,
    "appendGitTag": "--merge"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "pkg": {
    "scripts": [
      "node_modules/axios/dist/node/*",
      "dist/**/*.js"
    ],
    "assets": [
      "node_modules/figlet/fonts/*"
    ],
    "targets":[
      "win-x64"
    ],
    "outputPath": "executable/v%npm_package_version%/xxx"
  },
  "devDependencies": {
    "@faker-js/faker": "^8.4.1",
    "@types/activedirectory2": "^1.2.6",
    "@types/cli-progress": "^3.11.5",
    "@types/convert-array-to-csv": "^2.0.2",
    "@types/figlet": "^1.5.8",
    "@types/inquirer": "^9.0.7",
    "@types/jest": "^29.5.12",
    "@types/node": "^20.10.4",
    "@types/prompts": "^2.4.9",
    "@yao-pkg/pkg": "^5.11.5",
    "commander": "^11.1.0",
    "figlet": "^1.7.0",
    "jest": "^29.7.0",
    "rimraf": "^5.0.5",
    "ts-jest": "^29.1.2",
    "ts-node": "^10.9.2",
    "tsc-watch": "^6.0.4",
    "typescript": "^5.3.3"
  },
  "dependencies": {
    "activedirectory2": "link:node-activedirectory2-2.3.0",
    "axios": "^1.6.2",
    "axios-retry": "^4.0.0",
    "chalk": "^4.1.2",
    "cli-progress": "^3.12.0",
    "cli-table3": "^0.6.3",
    "convert-array-to-csv": "^2.0.0",
    "dotenv": "^16.3.1",
    "fast-xml-parser": "^4.3.2",
    "https-proxy-agent": "^7.0.2",
    "inquirer": "^8",
    "loading-cli": "^1.1.2",
    "prompts": "^2.4.2"
  }
}

```

And this is what I get in the terminal when I execute the “createExe” :

```shell
pnpm createExe

> xxx@1.9.0 createExe /Users/xxxx/Library/CloudStorage/xxxx/Git/xxxx
> pnpm build && pnpm rimraf executable/v%npm_package_version% && pkg .


> xxx@1.9.0 prebuild /Users/xxxxx/Library/CloudStorage/xxxx/Git/xxxx
> node -p "'export const VERSION = ' + JSON.stringify(require('./package.json').version) + ';'" > src/version.ts


> xxx@1.9.0 build /Users/xxx/Library/CloudStorage/xxx/Git/xxx
> pnpm rimraf dist/ && pnpm tsc

> pkg@5.11.5
> Fetching base Node.js binaries to PKG_CACHE_PATH
  fetched-v18.19.1-win-x64            [                    ] 0%> Not found in remote cache:
  {"tag":"v3.5","name":"node-v18.19.1-win-x64"}
> Building base binary from source:
  built-v18.19.1-win-x64
> Error! Not able to build for 'win' here, only for 'macos'
 ELIFECYCLE  Command failed with exit code 2.
```

**Reply by robertsLando** [reply]
When one of your dependencies is using a node native addon (`.node` lib).

Anyway the error is self explained

**Reply by Drilmo** [reply]
Forgive me if I don't understand it, but what am I supposed to understand? It's just not possible? If not, what should I do? Forgive me, but I'm at a loss. 
Thank you in any case for the time you take to reply and the help you can give me.

**Reply by robertsLando** [reply]
If your project doesn't have native addons you can do cross OS builds like compiling windows on mac.

About your error here the problem is that you didn't specified the target so it picks up your running nodejs version as target and there is no available pre-built nodejs patch so it tries to build one but you cannot as the patch for a specific os can be created only on that specific os,

To fix this try to specify node18-win-x64 in targets using `-t` options

**Reply by Drilmo** [reply]
However, I've specified the entire configuration in my package.json, in the pkg section, and it works fine when I change to linux or Mac, for example. And I don't understand, I always get an error even if I try to do it for linux.

**Reply by robertsLando** [reply]
I dunno, strange thing is that that patch is available here: https://github.com/yao-pkg/pkg-fetch/releases/tag/v3.5 (search for `node-v18.19.1-win-x64`) so it's strange that you get the error "cannot find in remote cache"

I would try to re-install pkg

**Reply by Drilmo** [reply, marked as answer]
Ok I found the reason for my mistake. I'm behind a proxy, and via my terminal I can't get the binary file associated with the packaging target from [https://github.com/yao-pkg/pkg-fetch/releases/tag/v3.5](https://github.com/yao-pkg/pkg-fetch/releases/tag/v3.5).

Once retrieved, I put it in “/Users/xxx/.pkg-cache/v3.5” then renamed it from “node-v18.19.1-win-x64” to “fetched-v18.19.1-win-x64”. 

Once done, I ran it again and it worked.
